### PR TITLE
HPCC-21929 Use QueryID in control:queries call if available

### PR DESCRIPTION
**Pull request by wangkx**
Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by wangkx** [comment]
@afishbeck please review.

File: esp/services/ws_workunits/ws_workunitsQuerySets.cpp [hunk 1]
```diff
@@ -1175,11 +1175,21 @@ IPropertyTree* getQueriesOnCluster(const char *target, const char *queryset, boo
 
     try
     {
+        StringBuffer control;
+        if (queryIDs.ordinality() == 0)
+            control.append("<control:queries/>");
+        else
+        {
+            control.append("<control:queries>");
+            for (unsigned i=0; i<queryIDs.ordinality(); i++)
```
Comment (afishbeck) [inline esp/services/ws_workunits/ws_workunitsQuerySets.cpp]: There is a macro for this: "ForEachItemIn(i, queryIDs)"

File: esp/services/ws_workunits/ws_workunitsQuerySets.cpp [hunk 2]
```diff
@@ -1158,7 +1158,7 @@ void retrieveQuerysetDetails(IEspContext &context, IArrayOf<IEspWUQuerySetDetail
     retrieveQuerysetDetails(context, details, registry, type, value, cluster, queriesOnCluster);
 }
 
-IPropertyTree* getQueriesOnCluster(const char *target, const char *queryset, bool checkAllNodes)
+IPropertyTree* getQueriesOnCluster(const char *target, const char *queryset, StringArray &queryIDs, bool checkAllNodes)
```
Comment (afishbeck) [inline esp/services/ws_workunits/ws_workunitsQuerySets.cpp]: Up to you, but if you make this "StringArray *queryIDs" then where you define emptyIDs below you could just use nullptr instead... and when you define queryIDs bellow you can pass by address "&queryIDs".

**Review by afishbeck — COMMENTED** [review]
@wangkx a couple trivial comments.

**Comment by wangkx** [comment]
@afishbeck please review my changes.

File: esp/services/ws_workunits/ws_workunitsQuerySets.cpp [hunk 3]
```diff
@@ -1176,13 +1176,13 @@ IPropertyTree* getQueriesOnCluster(const char *target, const char *queryset, Str
     try
     {
         StringBuffer control;
-        if (queryIDs.ordinality() == 0)
+        if (!queryIDs)
```
Comment (afishbeck) [inline esp/services/ws_workunits/ws_workunitsQuerySets.cpp]: I'd think !queryIDs or queryIDs->ordinality()==0 to make it easiest for callers.

**Review by afishbeck — CHANGES_REQUESTED** [review]
@wangkx one minor change.

**Comment by wangkx** [comment]
@afishbeck please review again.

**Review by afishbeck — APPROVED** [review]
@wangkx looks good.

**Comment by wangkx** [comment]
@richardkchapman this PR is ready to be merged.

**Pull request merged**